4. same as 2, but with implicit exception propagation and escalation/exampleUsage.js

```diff
@@ -0,0 +1,105 @@
+import * as userManager from './userManager.js'
+const { PERMISSIONS, UnauthorizedEx, NotFoundEx, MissingPropEx } = userManager
+
+function getUsername(userId, { userIdDoingRequest, fallback = null }) throws UnauthorizedEx {
+  try {
+    return userManager.getProperty(userId, 'username', { userIdDoingRequest })
+  } catch NotFoundEx {
+    return fallback
+  } catch MissingPropEx {
+    throw new Error('Unexpected failure to find the username property.')
+  }
+}
+
+function incrementAge(userId, { userIdDoingRequest, noopIfUserNotFound = false }) throws NotFoundEx {
+  try {
+    const age = userManager.getProperty(userId, 'age', { userIdDoingRequest })
+    userManager.setProperty(userId, 'age', age + 1, { userIdDoingRequest })
+  } catch NotFoundEx (ex) {
+    if (noopIfUserNotFound) return null
+    throw ex
+  } catch UnauthorizedEx {
+    throw new Error(`User with id ${userId} was not authorized to increment another's age. Please make sure they have the appropriate permissions before calling this function.`)
+  } catch MissingPropEx {
+    throw new Error('Unexpected failure to find the age property. please make sure the user has an age property before calling this function.')
+  }
+}
+
+function givePermissionToUsers(permission, userIds, { userIdDoingRequest }) throws UnauthorizedEx, NotFoundEx {
+  userManager.assertUserHasPermission(userIdDoingRequest, PERMISSIONS.view)
+  userManager.assertUserHasPermission(userIdDoingRequest, PERMISSIONS.update)
+
+  for (const userId of userIds) {
```

Comment:
There was actually a reason for two separate for loops in the previous versions - the idea was that we get any NotFound exceptions out of the way beforehand, that way we don't start making modifications to the users and stop half-way through because of an exception. This was supposed to demonstrate one of the types of places that an explicit-exceptions system would help prevent users from accidentally refactoring out an important ordering to some code :p

Reply:
Haha, actually this is quite a demonstration about why code dealing with exception could be very fragile! Didn't see that coming! :D

Actually when I was refactoring this part in my example, I was wondering why you were doing this operation in two separate loops in the first place. I even looked at your example `1. original idea`, in case I could find any reason for it, but still it didn't appeared obvious to me that it was for this specific reason. I guess that in this case, a comment explaining _why_ the code is structured like this would be much clearer than any syntax imaginable. 

For example, here, you wanted to have everything fail if at least one of the operations failed. But you could also be willing to have the same function that only fails invalid operations, while still completing the ones that didn't throw. So for a same function, you could implement different exception handling systems. The syntax could of course help to enforce certain patterns, but I think nothing will be clearer than a comment exactly explaining _why_ you wrote the code that way.

Reply:
hmm, so I guess that means both styles failed at their job. A simple comment certainly would have been very helpful there, like you said, especially since the double for-loop makes for some obvious, awkward code. And certainly, these explicit exceptions help prove the comment right - the comment just has to point out what can be seen by looking at the exceptions different functions are throwing (instead of saying "trust me, these earlier functions are throwing exception A").

In a recent project I've been working on, I've had to create a large number of functions that deal with exceptions in this intricate way - this is what inspired me to start pursuing different options to make this type of code more stable. In it, some of those functions are written with a wonky ordering in order to get certain exceptions out of the way before modifying things. Other functions are written with a normal-looking order - up-front it looks like other ordering would have worked just as well, but if someone actually tried to reorder it during some refactor, they would break things. My worry is that it's a lot of places that would need these comments, and it's easy for these comments to become outdated and confusing.

I guess the best option would be to just rely on unit tests. Someone comes along and tries to rearrange the ordering of things, and they break specific unit tests in the process.

I'll have to set and think on things some more. Maybe there's a completely different angle to all of this that would have removed a lot of these fragile points.